File: 数据结构篇/链表/206. 反转链表.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        def reverse(cur,prev):
            if not head:
                return prev
            tmp=cur.next
            cur.next=prev
            return reverse(cur=tmp,prev=cur)    
        
        return reverse(head,None)
        
        
        # 双指针
        # if head==None or head.next==None:
        #     return head 
        # cur=head
        # prev=None
        
        # while(cur):
        #     temp=cur.next
        #     cur.next=prev
        #     prev=cur
        #     cur=temp
            
        # return cur

File: 数据结构篇/链表/206.反转链表.cpp


#include <vector>
#include <queue>
#include <iostream>
using namespace std;

struct ListNode{
    int val;
    ListNode* next;
    ListNode(int x):val(x),next(NULL){}
};

class Solution
{
public:
ListNode* reverseList(ListNode* head)
{
    ListNode* temp;
    ListNode* cur=head;
    ListNode* pre=NULL;
    while(cur)
    {
        temp=cur->next;
        cur->next=pre;
        pre=cur;
        cur=temp;
    }
    return pre;
}
};

// 打印链表
void printList(ListNode* head) {
    while (head) {
        cout << head->val << " -> ";
        head = head->next;
    }
    cout << "NULL" << endl;
}

// 释放链表内存
void freeList(ListNode* head) {
    while (head) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

ListNode* createTestList(){
    ListNode* head= new ListNode(1);
    head->next= new ListNode(2);
    head->next->next= new ListNode(3);
    head->next->next->next= new ListNode(4);
    head->next->next->next->next= new ListNode(5);
    return head;


}

int main()
{
    ListNode* head=createTestList();
    Solution solution;
    ListNode* new_head=solution.reverseList(head);
    printList(new_head);
}